pkg/prometheus/server/operator.go

```diff
@@ -0,0 +1,2217 @@
+		}
+
+		stsReporter, err := operator.NewStatefulSetReporter(ctx, kclient, sset)
+		if err != nil {
+			return monitoringv1.PrometheusStatus{}, nil, errors.Wrapf(err, "failed to retrieve pods state for statefulset %s/%s", p.Namespace, ssetName)
+		}
+
+		res.Replicas += int32(len(stsReporter.Pods))
+		res.UpdatedReplicas += int32(len(stsReporter.UpdatedPods()))
+		res.AvailableReplicas += int32(len(stsReporter.ReadyPods()))
+		res.UnavailableReplicas += int32(len(stsReporter.Pods) - len(stsReporter.ReadyPods()))
+
+		for _, p := range stsReporter.Pods {
+			if p.Ready() && !stsReporter.IsUpdated(p) {
+				oldPods = append(oldPods, v1.Pod(*p))
+			}
+		}
+	}
+
+	return res, oldPods, nil
+}
+
+func (c *Operator) loadConfigFromSecret(sks *v1.SecretKeySelector, s *v1.SecretList) ([]byte, error) {
+	if sks == nil {
+		return nil, nil
+	}
+
+	for _, secret := range s.Items {
+		if secret.Name == sks.Name {
+			if c, ok := secret.Data[sks.Key]; ok {
+				return c, nil
+			}
+
+			return nil, fmt.Errorf("key %v could not be found in secret %v", sks.Key, sks.Name)
+		}
+	}
+
+	if sks.Optional == nil || !*sks.Optional {
+		return nil, fmt.Errorf("secret %v could not be found", sks.Name)
+	}
+
+	level.Debug(c.logger).Log("msg", fmt.Sprintf("secret %v could not be found", sks.Name))
+	return nil, nil
+}
+
+func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *monitoringv1.Prometheus, cg *prompkg.ConfigGenerator, ruleConfigMapNames []string, store *assets.Store) error {
+	// If no service or pod monitor selectors are configured, the user wants to
+	// manage configuration themselves. Do create an empty Secret if it doesn't
+	// exist.
+	if p.Spec.ServiceMonitorSelector == nil && p.Spec.PodMonitorSelector == nil &&
+		p.Spec.ProbeSelector == nil {
+		level.Debug(c.logger).Log("msg", "neither ServiceMonitor nor PodMonitor, nor Probe selector specified, leaving configuration unmanaged", "prometheus", p.Name, "namespace", p.Namespace)
+
+		s, err := prompkg.MakeEmptyConfigurationSecret(p, c.config)
+		if err != nil {
+			return errors.Wrap(err, "generating empty config secret failed")
+		}
+		sClient := c.kclient.CoreV1().Secrets(p.Namespace)
+		_, err = sClient.Get(ctx, s.Name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			if _, err := c.kclient.CoreV1().Secrets(p.Namespace).Create(ctx, s, metav1.CreateOptions{}); err != nil && !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "creating empty config file failed")
+			}
+		}
+		if !apierrors.IsNotFound(err) && err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	smons, err := c.selectServiceMonitors(ctx, p, store)
+	if err != nil {
+		return errors.Wrap(err, "selecting ServiceMonitors failed")
+	}
+
+	pmons, err := c.selectPodMonitors(ctx, p, store)
+	if err != nil {
+		return errors.Wrap(err, "selecting PodMonitors failed")
+	}
+
+	bmons, err := c.selectProbes(ctx, p, store)
+	if err != nil {
+		return errors.Wrap(err, "selecting Probes failed")
+	}
```

Comment:
`selectServiceMonitors`, `selectPodMonitors` and `selectProbes` will be identical for both Agent and Server, but they are methods from the Server operator. Here are some approaches that I've considered:

1. We could duplicate them for the agent operator, changing the argument from `monitoringv1.Prometheus` to `monitoringv1.PromethesAgent`. 
2. Another interface that will be shared between agent and server, but this time for their operators. `selectXXX` would receive the operator as an extra argument and we'd use `monitoringv1.PrometheusInterface` instead of `Prometheus` and `PrometheusAgent`

I have a feeling that the operator interface could reduce a lot of duplicated code, not only here, but I'm interested to know your thoughts!

Reply:
I'm always in favour of reducing code duplication 👍 However, it's hard for me to judge how much work would be in creating this new interface so my 2 cents would be to first follow 1. so we could focus on moving the agent forward and then we could do a follow-up PR implementing 2.

cc @simonpasquier for your thoughts

Reply:
@JoaoBraveCoding 

The rewrite was done in https://github.com/prometheus-operator/prometheus-operator/pull/5392, in case we go forward with that strategy :)